src/NServiceBus.Core/Unicast/MessageOperations.cs

```diff
@@ -155,7 +156,39 @@ Task ReplyMessage(IBehaviorContext context, Type messageType, object message, Re
 
             MergeDispatchProperties(outgoingContext, options.DispatchProperties);
 
-            return replyPipeline.Invoke(outgoingContext);
+            return InvokePipelineWithTracing(ActivityNames.OutgoingMessageActivityName, "reply", outgoingContext, replyPipeline);
+        }
+
+        static async Task InvokePipelineWithTracing<TContext>(string activityName, string displayName, TContext outgoingContext, IPipeline<TContext> pipeline)
+            where TContext : IBehaviorContext
+        {
+            using var activity = ActivitySources.Main.CreateActivity(activityName, ActivityKind.Producer);
+            if (activity == null)
+            {
+                await pipeline.Invoke(outgoingContext).ConfigureAwait(false);
+            }
+            else
+            {
+                activity.SetIdFormat(ActivityIdFormat.W3C);
+                activity.DisplayName = displayName;
+                activity.Start();
+                outgoingContext.Extensions.SetPipelineActitvity(activity);
+
+                try
+                {
+                    await pipeline.Invoke(outgoingContext).ConfigureAwait(false);
+                }
+#pragma warning disable PS0019 // When catching System.Exception, cancellation needs to be properly accounted for
+                catch (Exception ex)
+#pragma warning restore PS0019 // When catching System.Exception, cancellation needs to be properly accounted for
+                {
+                    // TODO: Add an explicit tag for operation canceled
+                    ActivityDecorator.SetErrorStatus(activity, ex);
+                    throw;
+                }
+                // TODO: should we stop the activity only once the message has been handed to the dispatcher?
+                activity.SetStatus(ActivityStatusCode.Ok);
```

Comment:
I wonder if this shouldn't be moved to `ActivityDecorator` to mirror what was done for `SetErrorStatus`. It means that both conditions end up at the same level of abstraction